scripts/esbuild/mock-doc.ts

```diff
@@ -0,0 +1,73 @@
+import type { BuildOptions as ESBuildOptions } from 'esbuild';
+import fs from 'fs-extra';
+import { join } from 'path';
+
+import { writeSizzleBundle } from '../bundles/plugins/sizzle-plugin';
+import { bundleDts } from '../utils/bundle-dts';
+import { BuildOptions } from '../utils/options';
+import { writePkgJson } from '../utils/write-pkg-json';
+import { getBaseEsbuildOptions, getEsbuildAliases, runBuilds } from './util';
+
+export async function buildMockDoc(opts: BuildOptions) {
+  const inputDir = join(opts.buildDir, 'mock-doc');
+  const srcDir = join(opts.srcDir, 'mock-doc');
+  const outputDir = opts.output.mockDocDir;
+
+  // bundle d.ts
+  await bundleMockDocDts(opts, inputDir, outputDir);
+
+  writePkgJson(opts, outputDir, {
+    name: '@stencil/core/mock-doc',
+    description: 'Mock window, document and DOM outside of a browser environment.',
+    main: 'index.cjs',
+    module: 'index.js',
+    types: 'index.d.ts',
+    sideEffects: false,
+  });
+
+  const mockDocAliases = getEsbuildAliases();
+
+  const sizzlePath = await writeSizzleBundle(opts);
+  mockDocAliases['sizzle'] = sizzlePath;
+
+  const mockDocBuildOptions: ESBuildOptions = {
+    ...getBaseEsbuildOptions(),
+    entryPoints: [join(srcDir, 'index.ts')],
+    bundle: true,
+    alias: mockDocAliases,
+    logLevel: 'info',
+    target: 'node16',
+  };
+
+  const esmOptions: ESBuildOptions = {
+    ...mockDocBuildOptions,
+    format: 'esm',
+    outfile: join(outputDir, 'index.js'),
+  };
+
+  const cjsOptions: ESBuildOptions = {
+    ...mockDocBuildOptions,
+    format: 'cjs',
+    outfile: join(outputDir, 'index.cjs'),
+  };
+
+  return runBuilds([esmOptions, cjsOptions], opts);
+}
+
+async function bundleMockDocDts(opts: BuildOptions, inputDir: string, outputDir: string) {
```

Comment:
This function seems quite different from the Rollup equivalent. How different are the written `index.d.ts` files? Also, was `index.d.ts-bundled.d.ts` file useless (IIUC, that isn't getting created anymore)?

Reply:
I did a bunch of manual checking and I believe that the files are the same, i.e. they end up exporting the same types. The function in the Rollup bundle ([here](https://github.com/ionic-team/stencil/blob/5e3605779589105d6a3da73fcfc2bbe5ceeb5def/scripts/bundles/mock-doc.ts#L79)) is basically a super manual way to bundle up and 'statically link' (sort of) a `.d.ts` file to make it standalone and portable, so it actually reads the files from disk, concatenates things, etc.

Here in the esbuild version I wanted to see if we could achieve the same result using our `bundleDts` function which leverages [dts-bundle-generator](https://github.com/timocov/dts-bundle-generator). Instead of being a somewhat 'homemade' solution this is a pretty reliable and tested library.

The reason for not writing a `-bundled.d.ts` version here (by passing `false` for the `useCache` parameter) is because I think for this code we always just want to generate a fresh new `index.d.ts` since it's not very expensive to do so and we can just avoid a whole potentially class of bugs that way

Reply:
Okay, that all makes sense. In that case, I think this is g2g!